### PR TITLE
[MULE-17907] - Fix mdc propagation when callback is called in a different thread.

### PR DESCRIPTION
**Pull request by fededonna**

### CODE REVIEW

**Comment by fededonna** [comment]
Mule issue created, linked with SE + new EE ticket to verify all services. Added tests, no changes with 1.4.x base branch.

File: src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java [hunk 1]
```diff
@@ -164,4 +204,46 @@ protected HttpRequestOptions getDefaultOptions(int responseTimeout) {
     return HttpRequestOptions.builder().responseTimeout(responseTimeout).build();
   }
 
+  private void testMdcPropagation(boolean shouldStream, boolean shouldThrowException) throws IOException {
+    String transactionId = UUID.getUUID();
+    Map<String, Object> capture = Maps.newHashMap();
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java]: move this down to right before the `try`

Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java]: why not use the old `new HashMap<>`?

File: src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java [hunk 2]
```diff
@@ -164,4 +204,46 @@ protected HttpRequestOptions getDefaultOptions(int responseTimeout) {
...
+                                      MDC.get("transactionId"));
+                          capture.put("currentThread", Thread
+                              .currentThread().getName());
+                        });
+
+      pollingProber.check(new ResponseReceivedProbe(responseReference));
+      assertThat(capture.get("exception"), shouldThrowException ? notNullValue() : nullValue());
+      assertThat(MDC.get("transactionId"), is(capture.get("transactionId")));
+      assertThat(MDC.get("currentThread"), is(not(capture.get("currentThread"))));
+      assertThat(responseReference.get().getStatusCode(), shouldThrowException ? is(500) : is(200));
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java]: Use the constants from `org.mule.runtime.http.api.HttpConstants.HttpStatus`

File: src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java [hunk 3]
```diff
@@ -164,4 +204,46 @@ protected HttpRequestOptions getDefaultOptions(int responseTimeout) {
...
+      // Release the lock on the streaming payload in order to finish without throwing an exception.
+      latch.release();
+    }
+    try {
+      client.sendAsync(shouldThrowException ? getRequest("http://localhost:9999") : getRequest(),
+                       getDefaultOptions(RESPONSE_TIMEOUT))
+          .whenComplete(
+                        (response, exception) -> {
+                          if (shouldThrowException) {
+                            responseReference.set(HttpResponse.builder().statusCode(500).build());
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java]: Use the constants from `org.mule.runtime.http.api.HttpConstants.HttpStatus`

File: src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java [hunk 4]
```diff
@@ -164,4 +204,46 @@ protected HttpRequestOptions getDefaultOptions(int responseTimeout) {
...
+                          } else {
+                            responseReference.set(response);
+                          }
+                          // since assertions won't fail here we need to capture this variables to assert later on.
+                          capture.put("exception", exception);
+                          capture.put("transactionId",
+                                      MDC.get("transactionId"));
+                          capture.put("currentThread", Thread
+                              .currentThread().getName());
```
Comment (elrodro83) [inline src/test/java/org/mule/service/http/impl/functional/client/HttpClientStreamingTestCase.java]: use static import for `currentThread()`. Check other occurrences

**Pull request closed without merging**